api/src/main/java/jakarta/json/Json.java

```diff
@@ -57,9 +63,57 @@
  * threads.
  */
 public final class Json {
-
+    private static final Map<ClassLoader, JsonProvider> providerCache = Collections.synchronizedMap(new WeakHashMap<>());
+    
     private Json() {
     }
+    
+    /**
+     * Sets the provider instance used by this factory class. The provider returned is associated with the context 
+     * class loader for the current thread (if one exists).
+     * 
+     * <p>Clients are encouraged to use this method to cache the desired provider instance, thus preventing 
+     * unnecessary ServiceLoader interactions (which may be costly).
+     * 
+     * @param provider a JsonProvider instance
+     */
+    public static void setProvider(JsonProvider provider) {
```

Comment:
AFAIK API changes are not allowed within context of Jakarta EE 9 but @m0mus knows that better, so he has to review this.

Reply:
No API changes please.

Reply:
@lukasj @m0mus I don't see how this could be accomplished *without* an API change. Please advise.

Reply:
I agree that should not change javax.json. 

What is the usecase/goal here?  I understand that finding the provider is expensive.  But the typical pattern is to create one of the factory (e.g. JsonParserFactory, JsonGeneratorFactory, JsonReaderFactory, JsonWriterFactory) once and reuse that.  So the cost of loading getting provider is typically amortized away.  Why do you also need to cache and reuse the provider?

Reply:
The issue is that the Json factory class is used extensively in other places (i.e. JsonPatch) without regards to provider caching which makes JsonPatch and JsonProvider expensive. This all happens internally and so this is a mechanism to force calls to Json factory class to use a cached provider if desired.

Reply:
I believe @m0mus is saying that we can't have API changes in Jakarta EE 9, but we should be able to have them for later releases, correct?

Reply:
@alexsm82 Correct, but `master` is for `Jakarta EE 9` release. If we want to use `master` for `Jakarta EE 10` release we need to do funny things with branches which is not good.

Reply:
Yep, understood. I just wanted to clarify for @JohnTimm that you weren't saying he needed to make this change without API updates, but rather that we can't have API updates for the next release. Should we create a branch for `Jakarta EE 10` that we can merge into master after the `Jakarta EE 9` release?

Reply:
@alexsm82 Yes we can if you not afraid rebasing.

Reply:
I was thinking about this PR. API changes are small and we can push it as part of Jakarta EE 9 release (although it's not recommended). We may have to be asked to go through another review process which (most probably) will be my responsibility to prepare for. I am not sure that it's worth to do it. 
Do you have any tests proving performance boost? Asking committers, do you folks think that we should merge it and take a burden of extra review process?

Reply:
@JohnTimm I'd like to see some performance details (scenarios and gains) to help determine if we want to push this for Jakarta EE 9.
@m0mus Do you have ruff details on the amount of effort the extra review process would be? Are there parts of this review process you could unload?  If so, I'll volunteer to help out.

Finally, should we update the TCK with new tests? I'm assuming this change will break signature tests at a minimum.

Reply:
@nmittles I'll find out about the additional review. I have no idea so far. Signature tests will be broken because of the package change anyway.

Reply:
I agree, I know we have had other complaints about this in the past, so it would be good to resolve, but it would be helpful to know the performance improvements and how much additional release work it would create to decide if it's worth targeting for Jakarta EE 9.

Reply:
There was a JsonPatch benchmark that @leadpony provided in the original issue. It would be nice to run that again with the new code to see where things are at. Otherwise, I will look into creating my own benchmark.